Source/WebCore/accessibility/AccessibilityNodeObject.cpp

```diff
@@ -1817,11 +1817,43 @@ void AccessibilityNodeObject::alternativeText(Vector<AccessibilityText>& textOrd
             textOrder.append(AccessibilityText(accessibleNameForNode(*object->node()), AccessibilityTextSource::Alternative));
     }
 
-    // The figure element derives its alternative text from the first associated figcaption element if one is available.
-    if (isFigureElement()) {
-        AccessibilityObject* captionForFigure = this->captionForFigure();
-        if (captionForFigure && !captionForFigure->isHidden())
-            textOrder.append(AccessibilityText(accessibleNameForNode(*captionForFigure->node()), AccessibilityTextSource::Alternative));
+    if (auto* image = dynamicDowncast<HTMLImageElement>(*node)) {
+        // https://github.com/w3c/aria/pull/2224
+        // Per html-aam, <img> elements that are unlabeled (e.g., alt attribute, ARIA, title) derive accname
+        // from an ancestor figure's <figcaption> if and only if the <figure> does not contain other flow content (besides the <figcaption>).
+        const AtomString& alt = image->attributeWithoutSynchronization(altAttr);
+
+        if (alt.isEmpty() && image->attributeWithoutSynchronization(titleAttr).isEmpty()) {
+            for (RefPtr ancestor = node->parentNode(); ancestor; ancestor = ancestor->parentNode()) {
+                if (auto* element = dynamicDowncast<HTMLElement>(ancestor.get()); element && element->hasTagName(figureTag)) {
+                    bool figureHasFlowContent = false;
+                    // Iterate over the direct children of the <img>'s ancestor <figure> for any common
+                    // flow content, including non-whitespace text nodes.
+                    for (RefPtr node = ancestor->firstChild(); node; node = node->nextSibling()) {
+                        if (auto* element = dynamicDowncast<HTMLElement>(node.get()); element && !element->hasTagName(figcaptionTag) && !element->hasTagName(imgTag)) {
+                            const AtomString& tag = element->localName();
+                            if (tag == blockquoteTag || tag == canvasTag || tag == codeTag || tag == divTag || tag == olTag || tag == pictureTag
+                            || tag == preTag || tag == pTag || tag == spanTag || tag == ulTag)
+                                figureHasFlowContent = true;
+                        }
+
+                        if (auto* text = dynamicDowncast<Text>(node.get()); text && !text->data().containsOnly<isASCIIWhitespace>())
+                            figureHasFlowContent = true;
+                    }
+                    // If no flow content is present in the <figure>, the <img> derives accname from its <figcaption>.
+                    if (!figureHasFlowContent) {
+                        RefPtr figureObject = objectCache ? objectCache->getOrCreate(*ancestor) : nullptr;
+                        RefPtr caption = figureObject && figureObject->isFigureElement() ? dynamicDowncast<AccessibilityNodeObject>(figureObject)->captionForFigure() : nullptr;
```

Comment:
My mistake as I realize this is exactly what I suggested. But if we aren't going to null-check the result of `dynamicDowncast` before dereferencing it, there's no point in using `dynamicDowncast` over `downcast`. Let's use `downcast` here.